compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala

```diff
@@ -211,10 +211,15 @@ trait PatternTypeConstrainer { self: TypeComparer =>
         tp
     }
 
+    val isWideningUnnecessary = patternTp match {
+      case tp @ AppliedType(tycon, args) =>
+        args.forall(_.isInstanceOf[TypeVar])
```

Comment:
Is there a need to check whether the type variables are instantiated?

Reply:
Ah, I guess that for type variables that are at all constrained, the subtyping doesn't have to hold.

Reply:
Just for clarity, I think that soon we will need to rewrite `constrainSimplePatternType` and inline some code from `isSubType`, so that we don't need to depend on either the skolem or the widening hacks. Having a flag to toggle widening is acceptable so that we avoid buggy behaviour in Scala 3.0.0, but I want to clean it up right afterwards. This PR is a small step in that direction.

Reply:
I was hoping that we'd be able to compile the snippet below with this PR, but sadly that's not the case. I'll need to look into it.
```
def foo(l: List[Int]) = l match { case _ : List[t] => 0 : t }
```